analyze: Treat TODOs as warnings instead of errors

IagoAbal opened the pull request:
This should allow data-flow analyses to work in the presence of
unsuported features---losing on precision or recall, of course.

IagoAbal commented:
I did small tests only and seems to work fine. Do we have some proper test that I could run? Maybe running taint analysis on some non-trivial Pyhton code or the like.

aryx commented on lang_GENERIC/analyze/AST_to_IL.ml:
```diff
@@ -71,16 +71,40 @@ let sgrep_construct any_generic =
 
+exception Todo of G.any
+
+(*s: function [[AST_to_IL.todo]] *)
```
Please do not modify or add those comments that start with s:, e:, or x:. They are added and used
by my literate programming tool syncweb, and modifying or adding new one will confure the tool.
Just ignore them. Even if you rename a function, do not modify the related comment.

IagoAbal replied:
Should I remove those `s:` and `e:` comments together with the previous definition of `todo`?

IagoAbal replied:
I removed the comments.

aryx commented on lang_GENERIC/analyze/AST_to_IL.ml:
```diff
@@ -71,16 +71,40 @@ let sgrep_construct any_generic =
+let todo_warning gany =
+  let toks = Lib_AST.ii_of_any gany in
+  let s = AST_generic.show_any gany in
+  let msg = spf "[DFA] Unsupported construct:\n<-----\n%s\n----->" s in
```
what does DFA mean?

IagoAbal replied:
DFA stands for Data-Flow Analysis. I actually wanted some feedback from you regarding the warnings. Right now these warnings will be rather verbose for a normal user, maybe the user should just get a brief "unsupported construct will affect precision/recall of dataflow analyses", and the verbose warning could be logged somewhere else (or in verbose mode). I haven't looked very careful into the different logging/verbosity options of Pfff.

IagoAbal replied:
I kept the warning message very simple, it simply informs about an unsupported construct and points to the source location. That should be enough to identify the construct in many cases. We could improve on that later on if needed.

aryx commented:
No there's no real testing code for tainting yet. Just a few tests in semgrep/semgrep-core/tests/OTHER/TAINTING/.
We could probably add something in semgrep/semgrep-core/tests/Test.ml to start some regression testing
on the tainting code like we do for semgrep (just run the engine on a test file and check that it reports an error (a flow from a tainting source to a sink) at the right line.

aryx commented:
Yes you can delete them.

IagoAbal commented:
> We could probably add something in semgrep/semgrep-core/tests/Test.ml to start some regression testing
> on the tainting code like we do for semgrep (just run the engine on a test file and check that it reports an error (a flow from a tainting source to a sink) at the right line.

I can take a look at it after this PR.

aryx commented:
The conflicts are due to @mjambon recent use of ocp-indent ...
You might need to restart from a fresh pfff repo (or try git merge develop).

IagoAbal committed:
analyze: Treat TODOs as warnings instead of errors

This should allow data-flow analyses to work in the presence of
unsupported features---losing on precision or recall, of course.

IagoAbal commented:
I "rebased" the changes semi-automatically. Not sure what are those checks that haven't completed yet...
